pkg/network/node/iptables_test.go

```diff
@@ -0,0 +1,53 @@
+package node
+
+import (
+	"fmt"
+	"testing"
+
+	"k8s.io/kubernetes/pkg/util/iptables"
+	iptablestest "k8s.io/kubernetes/pkg/util/iptables/testing"
+)
+
+func TestVxlanNoTrackRulesWithDefaultVxlanPort(t *testing.T) {
+	validateIPTableRuleForVxlanPort(t, 4789)
+}
+
+func TestVxlanNoTrackRulesWithCustomVxlanPort(t *testing.T) {
+	validateIPTableRuleForVxlanPort(t, 4788)
+}
+
+func validateIPTableRuleForVxlanPort(t *testing.T, dstPort uint32) {
+	ipt := iptablestest.NewFake()
+	nodeIpt := newNodeIPTables(ipt, nil, true, dstPort, uint32(0))
+	err := nodeIpt.syncIPTableRules()
+	if err != nil {
+		t.Fatalf("unexpected error while syncing ip table rules: %v", err)
+	}
+	exists, err := ipt.EnsureRule(iptables.Append, iptables.Table("raw"), iptables.Chain("PREROUTING"), "-m", "comment", "--comment", "disable conntrack for vxlan", "-j", "OPENSHIFT-NOTRACK")
```

Comment:
So i'm not sure `EnsureRule` Is the right function here.... At least for an instance of iptablestest it always seems to return `true` see -> https://github.com/kubernetes/kubernetes/blob/0fc6edeb1da8e3975be0cf5fc1b3f6467188de63/pkg/util/iptables/testing/fake.go#L102

What we want here is pretty much what the kubeproxy does, i.e https://github.com/kubernetes/kubernetes/blob/f3bf7e1ced956d5c23c5870699eae22eb55654c1/pkg/proxy/iptables/proxier_test.go#L4223

Where we let the `nodeIpt` sync and then ensure that the iptables rules are correctly programmed, I _think_ we can get all the programed iptables rules with [`FakeIPtables.save()`](https://github.com/kubernetes/kubernetes/blob/0fc6edeb1da8e3975be0cf5fc1b3f6467188de63/pkg/util/iptables/testing/fake.go#L122)

Please feel free to correct me if I'm missing something here.

Reply:
The k8s [FakeIPTables#Lines](https://github.com/kubernetes/kubernetes/blob/0fc6edeb1da8e3975be0cf5fc1b3f6467188de63/pkg/util/iptables/testing/fake.go#L61) is never used for populating rules and it does almost nothing for iptable methods that are needed by `nodeIpt.syncIPTableRules`, so save method can't be used. Looks like kube proxy has its own [in memory](https://github.com/kubernetes/kubernetes/blob/f3bf7e1ced956d5c23c5870699eae22eb55654c1/pkg/proxy/iptables/proxier.go#L225) which is then used for its tests validation and guess this fake iptables is for completion purpose. 

The new commit adds `NodeFakeIPTables` with necessary methods implemented for `nodeIpt.syncIPTableRules`, is that ok for now ?

Reply:
A little clunky... but that's alright since we don't maintain an internal Cache 

This is now functionally correct but my main NIT is that the test shouldn't be calling `ipt.ensureRule`(which is mutating ipt.Rules by creating the chain if not there) since that is being done by the node controller when  `err := nodeIpt.syncIPTableRules()` is called.  All the test should be doing is verifying that  `ipt.Rules` contains the correct rules, so maybe define `ipt.Present()` to just check for existence? 

Sorry to be a pain :)